src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,164 @@
+            newNode.prev = tail;
+            tail.next = newNode;
+            tail = newNode;
+        } else {
+            insertNode(findNode(index), newNode);
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T oldIem = findNode(index).value;
+        findNode(index).value = value;
+        return oldIem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        currentNode = findNode(index);
+        T removedItem = currentNode.value;
+        deleteItem(currentNode);
+        size--;
+        return removedItem;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        boolean isItemRemoved = false;
+        for (int i = 0; i < size; i++) {
+            currentNode = findNode(i);
+            if (currentNode.value == object
+                    || currentNode.value != null && currentNode.value.equals(object)) {
+                remove(i);
+                isItemRemoved = true;
+                break;
+            }
+        }
+        return isItemRemoved;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
+    }
+
+    private static class Node<E> {
+        private E value;
+        private Node<E> next;
+        private Node<E> prev;
+
+        Node(Node<E> prev, E value, Node<E> next) {
+            this.value = value;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private Node<T> findNode(int index) {
+        isInRangeIncludive(index);
+        if (index > (size / 2)) {
+            currentNode = tail;
+            int desiredIndex = size - 1;
+            while (index != desiredIndex) {
+                currentNode = currentNode.prev;
+                desiredIndex--;
+            }
+        } else {
+            currentNode = head;
+            int desiredIndex = 0;
+            while (index != desiredIndex) {
+                currentNode = currentNode.next;
+                desiredIndex++;
+            }
+        }
+        return currentNode;
+    }
+
+    private void isInRangeIncludive(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayIndexOutOfBoundsException("Invalid index");
+        }
+    }
+
+    private void isInRangeExcludive(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayIndexOutOfBoundsException("Invalid index");
+        }
+    }
+
+    private void setFirstItem(Node<T> newNode) {
+        head = newNode;
+        tail = newNode;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void deleteItem(Node<T> node) {
```

Comment:
it's better to rename it to unlink()
https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html